### ISSUE TITLE
Issues with fresh git-annex 7.20191114+git43-ge29663773-1~ndall+1

### ISSUE THREAD

**Issue by yarikoptic** [datalad/datalad#3890]
0.11.x  
https://travis-ci.org/datalad/datalad/builds/615882850?utm_source=github_status&utm_medium=notification
py 2.7 and 3.5 direct (so I guess upgrades to 7) 

```
FAIL: datalad.distribution.tests.test_create_sibling.test_target_ssh_simple
1146----------------------------------------------------------------------
1147Traceback (most recent call last):
1148  File "/home/travis/virtualenv/python2.7.15/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
1149    self.test(*self.arg)
1150  File "/home/travis/virtualenv/python2.7.15/lib/python2.7/site-packages/datalad/tests/utils.py", line 869, in newfunc
1151    return func(*args, **kwargs)
1152  File "/home/travis/virtualenv/python2.7.15/lib/python2.7/site-packages/datalad/tests/utils.py", line 921, in newfunc
1153    return func(*args, **kwargs)
1154  File "/home/travis/virtualenv/python2.7.15/lib/python2.7/site-packages/datalad/tests/utils.py", line 791, in newfunc
1155    t(*(arg + (uri,)), **kw)
1156  File "/home/travis/virtualenv/python2.7.15/lib/python2.7/site-packages/datalad/tests/utils.py", line 615, in newfunc
1157    return t(*(arg + (filename,)), **kw)
1158  File "/home/travis/virtualenv/python2.7.15/lib/python2.7/site-packages/datalad/tests/utils.py", line 615, in newfunc
1159    return t(*(arg + (filename,)), **kw)
1160  File "/home/travis/virtualenv/python2.7.15/lib/python2.7/site-packages/datalad/distribution/tests/test_create_sibling.py", line 332, in test_target_ssh_simple
1161    assert_set_equal(modified_files, ok_modified_files)
1162AssertionError: Items in the second set but not the first:
1163'.git/objects/info/packs'
1164'.git/info/refs'
1165
1166======================================================================
1167FAIL: datalad.support.tests.test_annexrepo.test_AnnexRepo_dirty
1168----------------------------------------------------------------------
1169Traceback (most recent call last):
1170  File "/home/travis/virtualenv/python2.7.15/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
1171    self.test(*self.arg)
1172  File "/home/travis/virtualenv/python2.7.15/lib/python2.7/site-packages/datalad/tests/utils.py", line 615, in newfunc
1173    return t(*(arg + (filename,)), **kw)
1174  File "/home/travis/virtualenv/python2.7.15/lib/python2.7/site-packages/datalad/support/tests/test_annexrepo.py", line 1854, in test_AnnexRepo_dirty
1175    ok_(not repo.dirty)
1176AssertionError: None
```
and for v5 (also I guess upgrades to 7):
```
======================================================================
ERROR: datalad.interface.tests.test_ls_webui.test_ls_json
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/virtualenv/python2.7.15/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/travis/virtualenv/python2.7.15/lib/python2.7/site-packages/datalad/tests/utils.py", line 442, in newfunc
    return t(*(arg + (d,)), **kw)
  File "/home/travis/virtualenv/python2.7.15/lib/python2.7/site-packages/datalad/tests/utils.py", line 554, in newfunc
    return tfunc(*(args + (path, url)), **kwargs)
  File "/home/travis/virtualenv/python2.7.15/lib/python2.7/site-packages/datalad/interface/tests/test_ls_webui.py", line 173, in test_ls_json
    ds.add(opj('dir', 'subgit'))                        # add the non-dataset git repo to annex
  File "/home/travis/virtualenv/python2.7.15/lib/python2.7/site-packages/datalad/distribution/dataset.py", line 527, in apply_func
    return f(**kwargs)
  File "/home/travis/virtualenv/python2.7.15/lib/python2.7/site-packages/datalad/interface/utils.py", line 499, in eval_func
    return return_func(generator_func)(*args, **kwargs)
  File "/home/travis/virtualenv/python2.7.15/lib/python2.7/site-packages/datalad/interface/utils.py", line 487, in return_func
    results = list(results)
  File "/home/travis/virtualenv/python2.7.15/lib/python2.7/site-packages/datalad/interface/utils.py", line 474, in generator_func
    msg="Command did not complete successfully")
IncompleteResultsError: Command did not complete successfully [{'status': u'error', 'orig_request': u'dir/subgit', 'message': u"'dir/subgit' already exists in the index\n", 'raw_input': True, 'parentds': u'/tmp/datalad_temp_tree_test_ls_jsonlkIb3R', 'action': u'add', 'path': u'/tmp/datalad_temp_tree_test_ls_jsonlkIb3R/dir/subgit', 'type': u'dataset', 'refds': u'/tmp/datalad_temp_tree_test_ls_jsonlkIb3R'}]
======================================================================
FAIL: datalad.distribution.tests.test_create_sibling.test_target_ssh_simple
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/virtualenv/python2.7.15/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/travis/virtualenv/python2.7.15/lib/python2.7/site-packages/datalad/tests/utils.py", line 869, in newfunc
    return func(*args, **kwargs)
  File "/home/travis/virtualenv/python2.7.15/lib/python2.7/site-packages/datalad/tests/utils.py", line 921, in newfunc
    return func(*args, **kwargs)
  File "/home/travis/virtualenv/python2.7.15/lib/python2.7/site-packages/datalad/tests/utils.py", line 791, in newfunc
    t(*(arg + (uri,)), **kw)
  File "/home/travis/virtualenv/python2.7.15/lib/python2.7/site-packages/datalad/tests/utils.py", line 615, in newfunc
    return t(*(arg + (filename,)), **kw)
  File "/home/travis/virtualenv/python2.7.15/lib/python2.7/site-packages/datalad/tests/utils.py", line 615, in newfunc
    return t(*(arg + (filename,)), **kw)
  File "/home/travis/virtualenv/python2.7.15/lib/python2.7/site-packages/datalad/distribution/tests/test_create_sibling.py", line 332, in test_target_ssh_simple
    assert_set_equal(modified_files, ok_modified_files)
AssertionError: Items in the second set but not the first:
'.git/objects/info/packs'
'.git/info/refs'
======================================================================
FAIL: datalad.metadata.tests.test_aggregation.test_partial_aggregation
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/virtualenv/python2.7.15/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/travis/virtualenv/python2.7.15/lib/python2.7/site-packages/datalad/tests/utils.py", line 442, in newfunc
    return t(*(arg + (d,)), **kw)
  File "/home/travis/virtualenv/python2.7.15/lib/python2.7/site-packages/datalad/metadata/tests/test_aggregation.py", line 403, in test_partial_aggregation
    ok_clean_git(path)
  File "/home/travis/virtualenv/python2.7.15/lib/python2.7/site-packages/datalad/tests/utils.py", line 211, in ok_clean_git
    eq_(index_diffs, [])
AssertionError: ['sub1\n=======================================================\nlhs: 160000 | 1f25e78d89a958ce5c0cdd0f31b6da24a6ca586a\nrhs: 160000 | 1f25e78d89a958ce5c0cdd0f31b6da24a6ca586a'] != []
```
master: https://travis-ci.org/datalad/datalad/builds/615883142?utm_source=github_status&utm_medium=notification  also has
```
======================================================================
ERROR: datalad.interface.tests.test_rerun.test_run_inputs_outputs
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/virtualenv/python3.5.6/lib/python3.5/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/travis/virtualenv/python3.5.6/lib/python3.5/site-packages/datalad/tests/utils.py", line 429, in newfunc
    return t(*(arg + (d,)), **kw)
  File "/home/travis/virtualenv/python3.5.6/lib/python3.5/site-packages/datalad/tests/utils.py", line 602, in newfunc
    return t(*(arg + (filename,)), **kw)
  File "/home/travis/virtualenv/python3.5.6/lib/python3.5/site-packages/datalad/interface/tests/test_rerun.py", line 613, in test_run_inputs_outputs
    Dataset(op.join(*((src,) + subds))).create(force=True).save()
  File "/home/travis/virtualenv/python3.5.6/lib/python3.5/site-packages/datalad/distribution/dataset.py", line 523, in apply_func
    return f(**kwargs)
  File "/home/travis/virtualenv/python3.5.6/lib/python3.5/site-packages/datalad/interface/utils.py", line 485, in eval_func
    return return_func(generator_func)(*args, **kwargs)
  File "/home/travis/virtualenv/python3.5.6/lib/python3.5/site-packages/datalad/interface/utils.py", line 473, in return_func
    results = list(results)
  File "/home/travis/virtualenv/python3.5.6/lib/python3.5/site-packages/datalad/interface/utils.py", line 460, in generator_func
    msg="Command did not complete successfully")
datalad.support.exceptions.IncompleteResultsError: Command did not complete successfully [{'status': 'error', 'action': 'add_submodule', 'path': '/tmp/datalad_temp_tree_test_run_inputs_outputssn2oicbi/s0/s1_0/s2', 'message': 'The following path is ignored by one of your .gitignore files:\ns1_0/s2\nUse -f if you really want to add it.\n', 'type': 'dataset'}, {'status': 'error', 'action': 'add_submodule', 'path': '/tmp/datalad_temp_tree_test_run_inputs_outputssn2oicbi/s0/s1_1/s2', 'message': 'The following path is ignored by one of your .gitignore files:\ns1_1/s2\nUse -f if you really want to add it.\n', 'type': 'dataset'}]
```
may be more ... didn't look into details yet.but there are changes in behavior for annex (and/or git?)

**Comment by kyleam** [comment]
I've started to look into some of these.

I can trigger the `test_AnnexRepo_dirty` failure *intermittently* on master (f0402869a) and 0.11.x (892ef795f) with a git-annex built from the latest commit (a5fe4d0ac).  The unexpected dirty repo is coming from file1.txt, a file tracked by git, being converted to an annex pointer file:

```
$ git diff
diff --git a/file1.txt b/file1.txt
index 339f0be..336e17d 100644
--- a/file1.txt
+++ b/file1.txt
@@ -1 +1 @@
-something else
\ No newline at end of file
+/annex/objects/SHA256E-s14--f41f3fa625ff120ddca7ef456bf66371ecea23c129f4e4c32367101edb516cf8.txt
```

Given the flakiness, I haven't tried to bisect it against the annex repo, but I'm able to trigger it with the latest annex release (7.20191114) too.  I've tried to reduce the DataLad script to a shell script that does the same operations, but file1.txt is left unmodified (the expected behavior) in the many times I've run the script.

The other tests I've tried locally:

* I've spent less time looking at `test_partial_aggregation`, but it also fails only intermittently on my end.

* I haven't been able to trigger the `test_ls_webui.py:test_ls_json` failure locally.

* `test_run_inputs_outputs` fails on my end reliably.  I've yet to look into it, but I plan on having a look at git version differences first.

**Comment by kyleam** [comment]
> * `test_run_inputs_outputs` fails on my end reliably.  I've yet to look into it, but I plan on having a look at git version differences first.

It boils down to a change in Git v2.24.0 that leads to `ls-files -o ...` traversing subdatasets.  I've [posted][0] to the Git mailing list about it.  Given that this involves the most recent Git release, let's see how things settle before considering how to deal with it on our end.

[0]: https://lore.kernel.org/git/87fti15agv.fsf@kyleam.com/T/#u

**Comment by kyleam** [comment]
Going through the last cron builds for [master][0] and [0.11.x][1], here's the list of failures I spot.

- [ ] datalad.interface.tests.test_ls_webui.test_ls_json
  maybe just 0.11.x?  As mentioned above, I haven't been able to trigger this locally.
- [ ] datalad.distribution.tests.test_create_sibling.test_target_ssh_simple
  0.11.x and master
- [ ] datalad.metadata.tests.test_aggregation.test_partial_aggregation
  0.11.x and master, flaky
- [ ] datalad.support.tests.test_annexrepo.test_AnnexRepo_dirty
  0.11.x and master, flaky
- [X] datalad.distribution.tests.test_install.test_install_noautoget_data
  master
- [X] datalad.core.local.tests.test_diff.test_path_diff
  master

The checked ones were papered over by gh-3904, with the underlying issue resolved upstream with [this in-flight series][2].

[0]: https://travis-ci.org/datalad/datalad/builds/621730974
[1]: https://travis-ci.org/datalad/datalad/builds/623302060
[2]: https://lore.kernel.org/git/pull.676.v2.git.git.1576008027.gitgitgadget@gmail.com/

**Comment by kyleam** [comment]
> I can trigger the `test_AnnexRepo_dirty` failure _intermittently_ on master ([f040286](https://github.com/datalad/datalad/commit/f0402869a95d453f76ce29ffbb63c0e5a75c0da0)) and 0.11.x ([892ef79](https://github.com/datalad/datalad/commit/892ef795f18fc633a433f59b3b616af69c6434a3)) with a git-annex built from the latest commit (a5fe4d0ac). [...]  I've tried to reduce the DataLad script to a shell script that does the same operations, but file1.txt is left unmodified (the expected behavior) in the many times I've run the script.

I revisited it and was able to come up with a script that triggered it on my end.  [Reported][0] on the git-annex site.

[0]: https://git-annex.branchable.com/bugs/A_case_where_file_tracked_by_git_unexpectedly_becomes_annex_pointer_file/

**Comment by kyleam** [comment]
> ======================================================================
> FAIL: datalad.distribution.tests.test_create_sibling.test_target_ssh_simple
>
>     assert_set_equal(modified_files, ok_modified_files)
> AssertionError: Items in the second set but not the first:
> '.git/objects/info/packs'
> '.git/info/refs'

Here's what's being tested there:

https://github.com/datalad/datalad/blob/c01fe8f98f15f252617a5902f3424f88e718adf9/datalad/distribution/tests/test_create_sibling.py#L317-L322

It seems very likely that the failure is due to using a later version of Git, but I'd guess the only problem that points to is that `test_target_ssh_simple` reaches much too far into the implementation guts of git.

**Comment by kyleam** [comment]
> * datalad.support.tests.test_annexrepo.test_AnnexRepo_dirty
>   0.11.x and master, flaky
>   [git-annex report](https://git-annex.branchable.com/bugs/A_case_where_file_tracked_by_git_unexpectedly_becomes_annex_pointer_file/)

This particular case has been fixed upstream in ea3cb7d27 (2019-12-27).  See Joey's comments in that bug report for why in general using `-c annex.largefiles=anything/nothing` may lead to accidental conversion.  Note that `git annex add` may gain `--git` and `--annex` flags, in which case we'll want to adjust `AnnexRepo._save_add` and friends to map `git=True,False` to those options (though to do so we'd of course need to either condition on the annex version or bump the minimum required version).

**Comment by kyleam** [comment]
OK, to recap the status indicated by the [check list][0], two of six failures remain unaddressed in any way.  Both of these tests appear flaky based on Travis runs.  One of these I've never triggered locally, and the other one I was able to trigger early on but not recently.  I don't have any good ideas how to troubleshoot those.  I'm going to open dedicated issues for them and then close this issue.

[0]: https://github.com/datalad/datalad/issues/3890#issuecomment-565027169
